Nexogen.Libraries.Metrics.Prometheus.AspCore/PrometheusExtensions.cs

```diff
@@ -26,14 +31,99 @@ public static IServiceCollection AddPrometheus(this IServiceCollection services)
         /// Add Prometheus Metrics support to the application.
         /// </summary>
         /// <param name="builder"></param>
+        /// <param name="getHttpPath">Function to retrieve metric path from http context</param>
         /// <returns></returns>
-        public static IApplicationBuilder UsePrometheus(this IApplicationBuilder builder)
+        public static IApplicationBuilder UsePrometheus(this IApplicationBuilder builder, System.Func<HttpContext, string> getHttpPath = null)
```

Comment:
@kodfodrasz agreed this was a bit of test induced damage. I'm not a big fan of inheritance instead preferring composition would be OK if I made the method taking the ```Func``` internal? and add an ```InternalsVisibleTo``` assembly attribute? I would still need to given we don't expose the middleware directly.

Reply:
Hi!

While I also generally prefer composition over inheritence, in this case I have a few reasons to still stick to inherritance.

1. There is no need to modify, extend the APIs, even with internal overloads.
2. This way there is no need for input validation logic in the implementation.  
    The line `this.pathBuilder = pathBuilder ?? PrometheusExtensions.GetHttpMetricPath;` (or its equivalent depending where the function finally lads after some refactoring) is basically to equivalent to: 
    `if(isTestMode){ ... } else { ... }` which can be totally saved without any input validation by using the feature of the laguage.
3. This way the subclass needed for tests can completely reside in the test project, which makes it cleaner and simpler to understand for future contributors.
4. `Func` is a very weak contract. Maybe because of past Java experiences, but I prefer single method interfaces in similiar cases, where the interface and the method name can be more descriptively named, documented. Funcs have their use also, but for more general usecases, no this specific.

I understand that InternalsVisible will still be needed. that is a compromise we have to make now anyway.

If you find these points rational may I ask you to please refactor to the subclass based test approach?

If you do not wish to do these changes, or do not have the capacity for them please tell us. We can accept the PR in that case, as your contribution is valuable. However in this case we may carry out the refactoring to better suite our organization's coding preferences as soon as we have capacity. In that case we will only release the next version with your contibution only after that.

Reply:
@kodfodrasz thanks for the feedback, your points are alright. In this particular scenario I don't have too much of a strong opinion and either way works; I'll implement the changes as you've requested. I'll make the changes and put in the PR. 

I'm using the library in all my projects going forward so it's only fair that I contribute :)

About 4: You can use a specific delegate type to achieve self documenting code. Delegates IMHO make code simpler because I don't need to define a class before I can consume it + any method that has a similar signature + return type can be used and I love that

Reply:
@gideonkorir Thank you very much, both for your understanding and your contribution.

I don't like delegates so much, but they have their benefits indeed. But *de gustibus non disputandum est*, as the ancient romans said.